File: folkmq/src/main/java/org/noear/folkmq/client/MqClientDefault.java
package org.noear.folkmq.client;

import org.noear.folkmq.common.MqConstants;
import org.noear.folkmq.exception.FolkmqException;
import org.noear.socketd.SocketD;
import org.noear.socketd.exception.SocketdAlarmException;
import org.noear.socketd.exception.SocketdConnectionException;
import org.noear.socketd.transport.client.Client;
import org.noear.socketd.transport.client.ClientConfigHandler;
import org.noear.socketd.transport.core.Entity;
import org.noear.socketd.transport.core.Session;
import org.noear.socketd.transport.core.entity.StringEntity;
import org.noear.socketd.transport.core.listener.EventListener;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.CompletableFuture;

/**
 * 消息客户端默认实现
 *
 * @author noear
 * @since 1.0
 */
public class MqClientDefault extends EventListener implements MqClientInternal {
    private static final Logger log = LoggerFactory.getLogger(MqClientDefault.class);

    //服务端地址
    private String serverUrl;
    //客户端
    private Client client;
    //客户端会话
    private Session clientSession;
    //客户端配置
    private ClientConfigHandler clientConfigHandler;
    //订阅字典
    private Map<String, MqSubscription> subscriptionMap = new HashMap<>();

    //自动回执
    private boolean autoAcknowledge = true;
    //发布重试
    private int publishRetryTimes = 0;

    public MqClientDefault(String serverUrl) {
        this.serverUrl = serverUrl.replace("folkmq://", "sd:tcp://");

        //接收派发指令
        on(MqConstants.MQ_EVENT_DISTRIBUTE, (s, m) -> {
            MqMessageReceivedImpl message = null;

            try {
                message = new MqMessageReceivedImpl(this, m);
                MqSubscription subscription = subscriptionMap.get(message.getTopic());

                if (subscription != null) {
                    subscription.consume(message);
                }

                //是否自动回执
                if (autoAcknowledge) {
                    acknowledge(message, true);
                }
            } catch (Throwable e) {
                if (message != null) {
                    acknowledge(message, false);
                    log.warn("Client consumer handle error, tid={}", message.getTid(), e);
                } else {
                    log.warn("Client consumer handle error", e);
                }
            }
        });
    }

    @Override
    public MqClient connect() throws IOException {
        client = SocketD.createClient(this.serverUrl);

        if (clientConfigHandler != null) {
            client.config(clientConfigHandler);
        }

        clientSession = client.listen(this).open();

        return this;
    }

    @Override
    public void disconnect() throws IOException {
        if (clientSession != null) {
            clientSession.close();
        }
    }

    @Override
    public MqClient config(ClientConfigHandler configHandler) {
        clientConfigHandler = configHandler;
        return this;
    }

    /**
     * 自动回执
     */
    @Override
    public MqClient autoAcknowledge(boolean auto) {
        this.autoAcknowledge = auto;
        return this;
    }

    @Override
    public MqClient publishRetryTimes(int times) {
        this.publishRetryTimes = times;
        return this;
    }

    /**
     * 订阅主题
     *
     * @param topic           主题
     * @param consumer        消费者（实例 ip 或 集群 name）
     * @param consumerHandler 消费处理
     */
    @Override
    public void subscribe(String topic, String consumer, MqConsumeHandler consumerHandler) throws IOException {
        MqSubscription subscription = new MqSubscription(topic, consumer, consumerHandler);

        //支持Qos1
        subscriptionMap.put(topic, subscription);

        if (clientSession != null && clientSession.isValid()) {
            Entity entity = new StringEntity("")
                    .meta(MqConstants.MQ_META_TOPIC, subscription.getTopic())
                    .meta(MqConstants.MQ_META_CONSUMER, subscription.getConsumer())
                    .at(MqConstants.BROKER_AT_SERVER_ALL);

            clientSession.sendAndRequest(MqConstants.MQ_EVENT_SUBSCRIBE, entity);

            log.info("Client subscribe successfully: {}#{}", topic, consumer);
        }
    }

    @Override
    public void unsubscribe(String topic, String consumer) throws IOException {
        subscriptionMap.remove(topic);

        if (clientSession != null && clientSession.isValid()) {
            Entity entity = new StringEntity("")
                    .meta(MqConstants.MQ_META_TOPIC, topic)
                    .meta(MqConstants.MQ_META_CONSUMER, consumer)
                    .at(MqConstants.BROKER_AT_SERVER_ALL);

            clientSession.sendAndRequest(MqConstants.MQ_EVENT_UNSUBSCRIBE, entity);

            log.info("Client unsubscribe successfully: {}#{}", topic, consumer);
        }
    }

    @Override
    public void publish(String topic, IMqMessage message) throws IOException {
        if (clientSession == null) {
            throw new SocketdConnectionException("Not connected!");
        }

        Entity entity = publishEntityBuild(topic, message);

        if (message.getQos() > 0) {
            //::Qos1
            if (publishRetryTimes > 0) {
                //采用同步 + 重试支持
                int times = publishRetryTimes;
                while (times > 0) {
                    try {
                        Entity resp = clientSession.sendAndRequest(MqConstants.MQ_EVENT_PUBLISH, entity);
                        int confirm = Integer.parseInt(resp.metaOrDefault(MqConstants.MQ_META_CONFIRM, "0"));
                        if (confirm == 1) {
                            break;
                        } else {
                            String messsage = "Client message publish confirm failed: " + resp.dataAsString();
                            throw new FolkmqException(messsage);
                        }
                    } catch (Throwable e) {
                        times--;
                        if (times == 0) {
                            throw e;
                        }
                    }
                }
            } else {

            }
        } else {
            //::Qos0
            clientSession.send(MqConstants.MQ_EVENT_PUBLISH, entity);
        }
    }

    /**
     * 发布消息
     *
     * @param topic   主题
     * @param message 消息
     */
    @Override
    public CompletableFuture<Boolean> publishAsync(String topic, IMqMessage message) throws IOException {
        if (clientSession == null) {
            throw new SocketdConnectionException("Not connected!");
        }

        CompletableFuture<Boolean> future = new CompletableFuture<>();
        Entity entity = publishEntityBuild(topic, message);

        if (message.getQos() > 0) {
            //采用异常 + 可选等待
            clientSession.sendAndRequest(MqConstants.MQ_EVENT_PUBLISH, entity, r -> {
                int confirm = Integer.parseInt(r.metaOrDefault(MqConstants.MQ_META_CONFIRM, "0"));
                if (confirm == 1) {
                    future.complete(true);
                } else {
                    String messsage = "Client message publish confirm failed: " + r.dataAsString();
                    future.completeExceptionally(new FolkmqException(messsage));
                }
            });
        } else {
            //::Qos0
            clientSession.send(MqConstants.MQ_EVENT_PUBLISH, entity);
            future.complete(true);
        }

        return future;
    }

    private Entity publishEntityBuild(String topic, IMqMessage message) {
        //构建消息实体
        StringEntity entity = new StringEntity(message.getContent());
        entity.meta(MqConstants.MQ_META_TID, message.getTid());
        entity.meta(MqConstants.MQ_META_TOPIC, topic);
        entity.meta(MqConstants.MQ_META_QOS, (message.getQos() == 0 ? "0" : "1"));
        if (message.getScheduled() == null) {
            entity.meta(MqConstants.MQ_META_SCHEDULED, "0");
        } else {
            entity.meta(MqConstants.MQ_META_SCHEDULED, String.valueOf(message.getScheduled().getTime()));
        }
        entity.at(MqConstants.BROKER_AT_SERVER);

        return entity;
    }

    /**
     * 消费回执
     *
     * @param message 收到的消息
     * @param isOk    回执
     */
    @Override
    public void acknowledge(MqMessageReceivedImpl message, boolean isOk) throws IOException {
        //发送“回执”，向服务端反馈消费情况
        if (message.getQos() > 0) {
            //此处用 replyEnd 不安全，时间长久可能会话断连过（流就无效了）
            clientSession.replyEnd(message.from, new StringEntity("")
                    .meta(MqConstants.MQ_META_ACK, isOk ? "1" : "0"));
        }
    }

    /**
     * 会话打开时
     */
    @Override
    public void onOpen(Session session) throws IOException {
        super.onOpen(session);

        log.info("Client session opened, sessionId={}", session.sessionId());

        //用于重连时重新订阅
        for (MqSubscription subscription : subscriptionMap.values()) {
            Entity entity = new StringEntity("")
                    .meta(MqConstants.MQ_META_TOPIC, subscription.getTopic())
                    .meta(MqConstants.MQ_META_CONSUMER, subscription.getConsumer())
                    .at(MqConstants.BROKER_AT_SERVER);

            session.send(MqConstants.MQ_EVENT_SUBSCRIBE, entity);
        }
    }

    /**
     * 会话关闭时
     */
    @Override
    public void onClose(Session session) {
        super.onClose(session);

        log.info("Client session closed, sessionId={}", session.sessionId());
    }

    /**
     * 会话出错时
     */
    @Override
    public void onError(Session session, Throwable error) {
        super.onError(session, error);

        if (log.isWarnEnabled()) {
            if (error instanceof SocketdAlarmException) {
                SocketdAlarmException alarmException = (SocketdAlarmException) error;
                log.warn("Client error, sessionId={}, from={}", session.sessionId(), alarmException.getFrom(), error);
            } else {
                log.warn("Client error, sessionId={}", session.sessionId(), error);
            }
        }
    }
}

File: folkmq-lang-python/folkmq/folkmq/client/MqClientDefault.py
from typing import Callable, Dict

from socketd import SocketD
from socketd.cluster.ClusterClientSession import ClusterClientSession
from socketd.exception.SocketDExecption import SocketDConnectionException, SocketDException
from socketd.transport.client.ClientConfig import ClientConfig
from socketd.transport.core import Entity
from socketd.transport.core.Message import Message
from socketd.transport.core.Session import Session
from socketd.transport.core.entity.EntityDefault import EntityDefault
from socketd.transport.core.entity.StringEntity import StringEntity
from socketd.transport.stream.RequestStream import RequestStream
from socketd.utils.LogConfig import log
from socketd.utils.RunUtils import RunUtils

from folkmq.client.MqAlarm import MqAlarm
from folkmq.client.MqClient import MqClientInternal
from folkmq.client.MqClientListener import MqClientListener
from folkmq.client.MqMessage import MqMessage
from folkmq.client.MqMessageReceived import MqMessageReceived, MqMessageReceivedImpl
from folkmq.client.MqSubscription import MqSubscription
from folkmq.client.MqTransaction import MqTransaction, MqTransactionImpl
from folkmq.common.MqAssert import MqAssert
from folkmq.common.MqConstants import MqConstants
from folkmq.common.MqMetasV2 import MqMetasV2
from folkmq.common.MqTopicHelper import MqTopicHelper
from folkmq.common.MqUtils import MqUtils
from folkmq.exception.FolkmqException import FolkmqException


# 消息客户端默认实现
class MqClientDefault(MqClientInternal):
    def __init__(self, urls:tuple[str], clientListener:MqClientListener|None = None):
        self._urls:tuple[str] = urls

        if clientListener is not None:
            self._clientListener = clientListener
        else:
            self._clientListener =  MqClientListener()

        self._clientListener.init(self)

        self._name:str|None = None
        self._namespace:str|None = None
        self._clientSession:ClusterClientSession|None = None
        self._clientConfigHandler:Callable[[ClientConfig], None]|None = None
        self._subscriptionMap:Dict[str, MqSubscription] = {}

        self._transactionCheckback:Callable[[MqMessageReceived],None] = None
        self._listenHandler:Callable[[MqMessageReceived],None] = None



    def name(self) -> str:
        return self._name

    def nameAs(self, name: str) -> 'MqClient':
        self._name = name
        return self

    def namespace(self) -> str:
        return self._namespace

    def namespaceAs(self, namespace: str) -> 'MqClient':
        self._namespace = namespace
        return self

    async def connect(self) -> 'MqClient':
        serverUrls:[str] = []

        for url in self._urls:
            url = url.replace("folkmq:ws://", "sd:ws://")
            url = url.replace("folkmq://", "sd:tcp://")
            serverUrls.append(url)

        self._clientSession = await (SocketD.create_cluster_client(*serverUrls)
            .config(self.__config_handle)
            .listen(self._clientListener)
            .open())

        return self

    def __config_handle(self, c:ClientConfig):
        from folkmq.FolkMQ import FolkMQ

        (c.meta_put(MqConstants.FOLKMQ_VERSION, FolkMQ.versionCodeAsString())
          .heartbeat_interval(6_000)
          .io_threads(1)
          .codec_threads(1)
          .exchange_threads(1))

        if self._name:
            c.meta_put("@", self._name)

        if self._namespace:
            c.meta_put(MqConstants.FOLKMQ_NAMESPACE, self._namespace)

        if self._clientConfigHandler:
            self._clientConfigHandler(c)

    async def disconnect(self):
        await self._clientSession.close()

    def config(self, configHandler: Callable[[ClientConfig], None]) -> 'MqClient':
        self._clientConfigHandler = configHandler
        return self

    def autoAcknowledge(self, auto: bool) -> 'MqClient':
        self._autoAcknowledge = auto
        return self

    async def subscribe(self, topic: str, consumerGroup: str | None, autoAck: bool | None,
                  consumerHandler: Callable[[MqMessageReceived], None]):
        if consumerGroup is None:
            consumerGroup = self.name()

        if autoAck is None:
            autoAck = self._autoAcknowledge

        MqAssert.requireNonNull(topic, "Param 'topic' can't be null")
        MqAssert.requireNonNull(consumerGroup, "Param 'consumerGroup' can't be null")
        MqAssert.requireNonNull(consumerHandler, "Param 'consumerHandler' can't be null")

        MqAssert.assertMeta(topic, "topic")
        MqAssert.assertMeta(consumerGroup, "consumerGroup")

        #支持命名空间
        topic = MqTopicHelper.getFullTopic(self._namespace, topic)

        subscription = MqSubscription(topic, consumerGroup, autoAck, consumerHandler)
        self._subscriptionMap[subscription.getQueueName()] = subscription

        if self._clientSession is not None:
            for session in self._clientSession.get_session_all():
                #如果有连接会话，则执行订阅
                entity = (StringEntity("")
                          .meta_put(MqConstants.MQ_META_TOPIC, subscription.getTopic())
                          .meta_put(MqConstants.MQ_META_CONSUMER_GROUP, subscription.getConsumerGroup())
                          .at(MqConstants.BROKER_AT_SERVER_ALL))
                #使用 Qos1
                await session.send_and_request(MqConstants.MQ_EVENT_SUBSCRIBE, entity, 30_000).waiter()
                log.info(f"Client subscribe successfully: {topic}#{consumerGroup}, sessionId={session.session_id()}")

    async def unsubscribe(self, topic: str, consumerGroup: str | None):
        if consumerGroup is None:
            consumerGroup = self.name()

        MqAssert.requireNonNull(topic, "Param 'topic' can't be null")
        MqAssert.requireNonNull(consumerGroup, "Param 'consumerGroup' can't be null")

        MqAssert.assertMeta(topic, "topic")
        MqAssert.assertMeta(consumerGroup, "consumerGroup")

        #支持命名空间
        topic = MqTopicHelper.getFullTopic(self._namespace, topic)

        queueName = topic + MqConstants.SEPARATOR_TOPIC_CONSUMER_GROUP + consumerGroup
        self._subscriptionMap.pop(queueName)

        if self._clientSession is not None:
            for session in self._clientSession.get_session_all():
                #如果有连接会话，则执行订阅
                entity = (StringEntity("")
                          .meta_put(MqConstants.MQ_META_TOPIC, topic)
                          .meta_put(MqConstants.MQ_META_CONSUMER_GROUP, consumerGroup)
                          .at(MqConstants.BROKER_AT_SERVER_ALL))
                #使用 Qos1
                await session.send_and_request(MqConstants.MQ_EVENT_UNSUBSCRIBE, entity, 30_000).waiter()
                log.info(f"Client unsubscribe successfully: {topic}#{consumerGroup}， sessionId={session.session_id()}")

    async def publish(self, topic: str, message: MqMessage):
        MqAssert.requireNonNull(topic, "Param 'topic' can't be null")
        MqAssert.requireNonNull(message, "Param 'message' can't be null")

        MqAssert.assertMeta(topic, "topic")

        if self._clientSession is None:
            raise SocketDConnectionException("Not connected!")

        # 支持命名空间
        topic = MqTopicHelper.getFullTopic(self._namespace, topic)
        session = self._clientSession.get_session_any(self._diversionOrNull(topic, message))

        if session is None or session.is_valid() == False:
            raise SocketDException("No session is available!")

        entity = MqUtils.getOf(session).publishEntityBuild(topic, message)

        if message.getQos() > 0:
            resp = await session.send_and_request(MqConstants.MQ_EVENT_PUBLISH, entity).waiter()
            confirm = int(resp.meta_or_default(MqConstants.MQ_META_CONFIRM, "0"))
            if confirm != 1:
                messsage = "Client message publish confirm failed: " + resp.data_as_string()
                raise FolkmqException(messsage)
        else:
            session.send(MqConstants.MQ_EVENT_PUBLISH, entity)

    async def unpublish(self, topic: str, key: str):
        MqAssert.requireNonNull(topic, "Param 'topic' can't be null")
        MqAssert.requireNonNull(key, "Param 'key' can't be null")

        MqAssert.assertMeta(topic, "topic")
        MqAssert.assertMeta(key, "key")

        if self._clientSession is None:
            raise SocketDConnectionException("Not connected!")

        #支持命名空间
        topic = MqTopicHelper.getFullTopic(self._namespace, topic)
        session = self._clientSession.get_session_any(None)
        if session is None or session.is_valid() == False:
            raise SocketDException("No session is available!")

        entity = (StringEntity("")
                  .meta_put(MqConstants.MQ_META_TOPIC, topic)
                  .meta_put(MqConstants.MQ_META_KEY, key)
                  .at(MqConstants.BROKER_AT_SERVER_ALL))

        resp = await session.send_and_request(MqConstants.MQ_EVENT_UNPUBLISH, entity).waiter()
        confirm = int(resp.meta_or_default(MqConstants.MQ_META_CONFIRM, "0"))
        if confirm != 1:
            messsage = "Client message unpublish confirm failed: " + resp.data_as_string()
            raise FolkmqException(messsage)



    async def listen(self, listenHandler: Callable[[MqMessageReceived], None]):
        # 检查必要条件
        if self._name is None:
            return FolkmqException("Client 'name' can't be empty")

        self._listenHandler = listenHandler

    def send(self, message: MqMessage, toName: str, timeout: int | None = None) -> RequestStream | None:
        # 检查必要条件
        if self._name is None:
            return FolkmqException("Client 'name' can't be empty")

        # 检查参数
        MqAssert.requireNonNull(toName, "Param 'toName' can't be null")
        MqAssert.requireNonNull(message, "Param 'message' can't be null")

        MqAssert.assertMeta(toName, "toName")

        if self._clientSession is None:
            raise SocketDConnectionException("Not connected!")

        session = self._clientSession.get_session_any(None)
        if session is None or session.is_valid() == False:
            raise SocketDException("No session is available!")

        message.internalSender(self.name())
        entity = MqUtils.getOf(session).publishEntityBuild("", message)
        entity.put_meta(MqMetasV2.MQ_META_CONSUMER_GROUP, toName)
        entity.at(toName)

        if message.getQos() > 0:
            return session.send_and_request(MqConstants.MQ_EVENT_REQUEST, entity, timeout)
        else:
            session.send(MqConstants.MQ_EVENT_REQUEST, entity)
            return None

    def transactionCheckback(self, transactionCheckback: Callable[[MqMessageReceived], None]):
        if transactionCheckback is not None:
            self._transactionCheckback = transactionCheckback
        return self

    def newTransaction(self) -> MqTransaction:
        if self._name is None:
            return FolkmqException("Client 'name' can't be empty")

        return MqTransactionImpl(self)

    async def publish2(self, tmid: str, keyAry: [str], isRollback: bool):
        if keyAry is None or len(keyAry) == 0:
            return

        if self._clientSession is None:
            raise SocketDConnectionException("Not connected!")

        session = self._clientSession.get_session_any(tmid)

        if session is None or session.is_valid() == False:
            raise SocketDException("No session is available!")

        entity = (StringEntity(",".join(keyAry))
                  .meta_put(MqConstants.MQ_META_ROLLBACK, ( "1" if isRollback else "0"))
                  .at(MqConstants.BROKER_AT_SERVER_HASH)) # 事务走哈希

        resp = await session.send_and_request(MqConstants.MQ_EVENT_PUBLISH2, entity).waiter()

        confirm = int(resp.meta_or_default(MqConstants.MQ_META_CONFIRM, "0"))
        if confirm != 1:
            messsage = "Client message publish2 confirm failed: " + resp.data_as_string()
            raise FolkmqException(messsage)


    def reply(self, session: Session, f: Message, message: MqMessageReceivedImpl, isOk: bool, entity: Entity):
        """发送“回执”，向服务端反馈消费情况"""
        if message.getQos() > 0:
            if session.is_valid():
                if entity is None:
                    entity = EntityDefault()

                if isinstance(entity, MqAlarm):
                    RunUtils.taskTry(session.send_alarm(f, entity.data_as_string()))
                else:
                    entity.put_meta(MqConstants.MQ_META_ACK, "1" if isOk else "0")
                    RunUtils.taskTry(session.reply_end(f, entity))

    def _diversionOrNull(self, fullTopic: str, message: MqMessage) -> str | None:
        if message.isTransaction():
            return message.getTmid()
        elif message.isSequence():
            if message.getSequenceSharding():
                return message.getSequenceSharding()
            else:
                return fullTopic
        else:
            return None

    def getSubscription(self, fullTopic: str, consumerGroup: str):
        queueName = fullTopic + MqConstants.SEPARATOR_TOPIC_CONSUMER_GROUP + consumerGroup
        return self._subscriptionMap.get(queueName)

    def getSubscriptionAll(self) -> [MqSubscription]:
        return self._subscriptionMap.values()
    def getSubscriptionSize(self):
        return len(self._subscriptionMap)